internal/controller/storageclient_controller.go

```diff
@@ -55,54 +64,143 @@ const (
-	storageClientFinalizer             = "storageclient.ocs.openshift.io"
-	storageClaimProcessedAnnotationKey = "ocs.openshift.io/storageclaim.processed"
-	storageClientDefaultAnnotationKey  = "ocs.openshift.io/storageclient.default"
+	storageClientNameLabel = "ocs.openshift.io/storageclient.name"
+	storageClientFinalizer = "storageclient.ocs.openshift.io"
 
 	// indexes for caching
-	ownerIndexName = "index:ownerUID"
+	pvClusterIDIndexName   = "index:persistentVolumeClusterID"
+	vscClusterIDIndexName  = "index:volumeSnapshotContentCSIDriver"
+	vgscClusterIDIndexName = "index:volumeGroupSnapshotContentCSIDriver"
+	ownerUIDIndexName      = "index:ownerUID"
+
+	VolumeGroupSnapshotClassCrdName = "volumegroupsnapshotclasses.groupsnapshot.storage.k8s.io"
 
 	csvPrefix = "ocs-client-operator"
 )
 
 // StorageClientReconciler reconciles a StorageClient object
 type StorageClientReconciler struct {
-	ctx context.Context
 	client.Client
-	Log           klog.Logger
-	Scheme        *runtime.Scheme
-	recorder      *utils.EventReporter
-	storageClient *v1alpha1.StorageClient
-
+	Scheme            *runtime.Scheme
 	OperatorNamespace string
+	AvailableCrds     map[string]bool
+
+	log               klog.Logger
+	ctx               context.Context
+	recorder          *utils.EventReporter
+	storageClient     *v1alpha1.StorageClient
+	storageClientHash string
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *StorageClientReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	ctx := context.Background()
-	if err := mgr.GetCache().IndexField(ctx, &v1alpha1.StorageClaim{}, ownerIndexName, func(obj client.Object) []string {
+	csiDrivers := []string{templates.RBDDriverName, templates.CephFsDriverName}
+	if err := mgr.GetCache().IndexField(ctx, &corev1.PersistentVolume{}, pvClusterIDIndexName, func(o client.Object) []string {
+		pv := o.(*corev1.PersistentVolume)
+		if pv != nil &&
+			pv.Spec.CSI != nil &&
+			slices.Contains(csiDrivers, pv.Spec.CSI.Driver) &&
+			pv.Spec.CSI.VolumeAttributes["clusterID"] != "" {
+			return []string{pv.Spec.CSI.VolumeAttributes["clusterID"]}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer for PV cluster id: %v", err)
+	}
+	if err := mgr.GetCache().IndexField(ctx, &snapapi.VolumeSnapshotContent{}, vscClusterIDIndexName, func(o client.Object) []string {
+		vsc := o.(*snapapi.VolumeSnapshotContent)
+		if vsc != nil &&
+			slices.Contains(csiDrivers, vsc.Spec.Driver) &&
+			vsc.Status != nil &&
+			vsc.Status.SnapshotHandle != nil {
+			parts := strings.Split(*vsc.Status.SnapshotHandle, "-")
+			if len(parts) == 9 {
+				// second entry in the volumeID is clusterID which is unique across the cluster
+				return []string{parts[2]}
+			}
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to set up FieldIndexer for VSC csi driver name: %v", err)
+	}
+
+	if r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
```

Comment:
Why a map? is there a reason to introduce a map and not just read/list from the cached client?

Reply:
`VolumeGroupSnapshotClassCrd` are not available in some OCP versions. @Madhu-1 do you know from which version these APIs are available by default?

Reply:
I checked w/ Rakshith, the API is not available by default in 4.19.

Reply:
sorry i missed the ping, Openshift is planning to have volumegroupsnapshot GA in next OCP 4.18.z release which means it might come in OCP 4.19